### PR TITLE
ASR2 CHiME4 & Gigaspeech Recipes

### PR DESCRIPTION
**Pull request by yichen14**
## What?

Add ASR discrete token recipes for CHiME4 and Gigaspeech.

## Why?

Extend ASR2 for both dataset, and update exp result.

## See also

None

### CODE REVIEW

File: egs2/TEMPLATE/asr2/asr2.sh [hunk 1]
```diff
@@ -37,7 +37,7 @@ ngpu=1               # The number of gpus ("0" uses cpu, otherwise use gpu).
 num_nodes=1          # The number of nodes.
 nj=32                # The number of parallel jobs.
 inference_nj=32      # The number of parallel jobs in decoding.
-gpu_inference=false  # Whether to perform gpu decoding.
+gpu_inference=       # Whether to perform gpu decoding.
```
Comment (simpleoier) [inline egs2/TEMPLATE/asr2/asr2.sh:40]: You may not need to change this.

File: egs2/TEMPLATE/asr2/asr2.sh [hunk 2]
```diff
@@ -649,13 +649,15 @@ if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ] && ! [[ " ${skip_stages} " =~ [
...
             _opts=
-            if [ -e "data/${dset}"/segments ]; then
+            if [ -e data/"${dset}"/segments ]; then
                 # "segments" is used for splitting wav files which are written in "wav".scp
                 # into utterances. The file format of segments:
                 #   <segment_id> <record_id> <start_time> <end_time>
                 #   "e.g. call-861225-A-0050-0065 call-861225-A 5.0 6.5"
                 # Where the time is written in seconds.
-                _opts+="--segments data/${dset}/segments "
+                # _opts+="--segments data/${dset}/segments "
```
Comment (simpleoier) [inline egs2/TEMPLATE/asr2/asr2.sh:658]: ditto. You may not need to change this.

File: egs2/TEMPLATE/asr2/asr2.sh [hunk 3]
```diff
@@ -680,6 +682,12 @@ fi
 if [ ${stage} -le 4 ] && [ ${stop_stage} -ge 4 ] && ! [[ " ${skip_stages} " =~ [[:space:]]4[[:space:]] ]]; then
     log "Stage 4a: Perform Kmeans using ${kmeans_feature_type} features"
 
+    if [[ "${audio_format}" == *ark* ]]; then
```
Comment (simpleoier) [inline egs2/TEMPLATE/asr2/asr2.sh:685]: ditto.

File: egs2/TEMPLATE/asr2/asr2.sh [hunk 4]
```diff
@@ -703,6 +711,8 @@ if [ ${stage} -le 4 ] && [ ${stop_stage} -ge 4 ] && ! [[ " ${skip_stages} " =~ [
 
     log "Stage 4b: Prepare token_list and convert number indices to CJK tokens"
 
+
```
Comment (simpleoier) [inline egs2/TEMPLATE/asr2/asr2.sh:714]: ditto.

File: egs2/chime4/asr2/README.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+- date: `Mon Sep 11 04:53:07 EDT 2023`
+- python version: `3.10.10 (main, Mar 21 2023, 18:45:11) [GCC 11.2.0]`
+- espnet version: `espnet 202308`
+- pytorch version: `pytorch 1.13.1+cu117`
+- Git hash: `83e687f3b41310a000f4a5b65857734709752bf6`
+  - Commit date: `Tue Aug 15 18:31:02 2023 -0400`
+
+## exp/asr_train_discrete_asr_e_branchformer_e12_mlp1024_linear1024_macaron_lr1e-4_warmup25k_conv1d1_raw_wavlm_large_21_km1000_bpe_rm2000_char_ts_sp
+
+- ASR config: conf/tuning/train_discrete_asr_e_branchformer_e12_mlp1024_linear1024_macaron_lr1e-4_warmup25k_conv1d1.yaml
```
Comment (simpleoier) [inline egs2/chime4/asr2/README.md:31]: you can add a file link to the config file.

File: egs2/chime4/asr2/README.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+- decode config: conf/decode_asr.yaml
```
Comment (simpleoier) [inline egs2/chime4/asr2/README.md:32]: ditto, a link to the file.

File: egs2/chime4/asr2/cmd.sh [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+# If jobs failed, your configuration might be wrong for your environment.
+#
+#
+# The official documentation for run.pl, queue.pl, slurm.pl, and ssh.pl:
+#   "Parallelization in Kaldi": http://kaldi-asr.org/doc/queue.html
+# =========================================================~
+
+
+# Select the backend used by run.sh from "local", "stdout", "sge", "slurm", or "ssh"
+cmd_backend='slurm'
```
Comment (simpleoier) [inline egs2/chime4/asr2/cmd.sh:31]: please use the local in the recipe.

File: egs2/chime4/asr2/conf/tuning/train_discrete_asr_e_branchformer_e12_mlp1024_linear1024_macaron_lr1e-4_warmup25k_conv1d1.yaml [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    rel_pos_type: latest
+    cgmlp_linear_units: 1024
+    cgmlp_conv_kernel: 31
+    use_linear_after_conv: false
+    gate_activation: identity
+    num_blocks: 12
+    dropout_rate: 0.1
+    positional_dropout_rate: 0.1
+    attention_dropout_rate: 0.1
+    input_layer: conv1d1
```
Comment (simpleoier) [inline egs2/chime4/asr2/conf/tuning/train_discrete_asr_e_branchformer_e12_mlp1024_linear1024_macaron_lr1e-4_warmup25k_conv1d1.yaml:22]: Can you add a comment why conv1d1 is used here?

File: egs2/chime4/asr2/local/data.sh
Comment (simpleoier) [inline egs2/chime4/asr2/local/data.sh:1]: you can link the `asr1/local/data.sh`

File: egs2/chime4/asr2/local/sym_channel.py
Comment (simpleoier) [inline egs2/chime4/asr2/local/sym_channel.py:1]: you can symlink to `asr1/local/sym_channel.py`.

File: egs2/gigaspeech/asr2/README.md
Comment (simpleoier) [inline egs2/gigaspeech/asr2/README.md:1]: Please clean this README file manually. Some results are not complete and not necessary. Add config link.

File: egs2/gigaspeech/asr2/cmd.sh [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+# If jobs failed, your configuration might be wrong for your environment.
+#
+#
+# The official documentation for run.pl, queue.pl, slurm.pl, and ssh.pl:
+#   "Parallelization in Kaldi": http://kaldi-asr.org/doc/queue.html
+# =========================================================~
+
+
+# Select the backend used by run.sh from "local", "stdout", "sge", "slurm", or "ssh"
+cmd_backend='slurm'
```
Comment (simpleoier) [inline egs2/gigaspeech/asr2/cmd.sh:31]: use `local` here.

File: egs2/gigaspeech/asr2/cmd.sh [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+# "sbatch" (Slurm)
+elif [ "${cmd_backend}" = slurm ]; then
+    # The default setting is written in conf/slurm.conf.
+    # You must change "-p cpu" and "-p gpu" for the "partition" for your environment.
+    # To know the "partion" names, type "sinfo".
+    # You can use "--gpu * " by default for slurm and it is interpreted as "--gres gpu:*"
+    # The devices are allocated exclusively using "${CUDA_VISIBLE_DEVICES}".
+
+    export train_cmd="slurm.pl --num_threads 2"
```
Comment (simpleoier) [inline egs2/gigaspeech/asr2/cmd.sh:83]: Please remove your customized configs.

File: egs2/gigaspeech/asr2/conf/slurm.conf
Comment (simpleoier) [inline egs2/gigaspeech/asr2/conf/slurm.conf:1]: please remove your customized configs.

File: egs2/gigaspeech/asr2/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml
Comment (simpleoier) [inline egs2/gigaspeech/asr2/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml:1]: Did you have the conformer results? If not, you may not need to include this file.

File: egs2/gigaspeech/asr2/conf/tuning/train_discrete_asr_e_branchformer1_e12.yaml
Comment (simpleoier) [inline egs2/gigaspeech/asr2/conf/tuning/train_discrete_asr_e_branchformer1_e12.yaml:1]: Did you have this config? If not, you may not need to include this file.

File: egs2/gigaspeech/asr2/conf/tuning/train_discrete_asr_e_branchformer1_e12_subsamp.yaml
Comment (simpleoier) [inline egs2/gigaspeech/asr2/conf/tuning/train_discrete_asr_e_branchformer1_e12_subsamp.yaml:1]: ditto.

File: egs2/gigaspeech/asr2/conf/tuning/train_discrete_asr_e_branchformer1_e17_size512_mlp3072_linear1024_layerdrop.yaml
Comment (simpleoier) [inline egs2/gigaspeech/asr2/conf/tuning/train_discrete_asr_e_branchformer1_e17_size512_mlp3072_linear1024_layerdrop.yaml:1]: ditto.

File: egs2/gigaspeech/asr2/local/data.sh
Comment (simpleoier) [inline egs2/gigaspeech/asr2/local/data.sh:1]: You can symlink to asr1 files.

File: egs2/gigaspeech/asr2/local/score.sh
Comment (simpleoier) [inline egs2/gigaspeech/asr2/local/score.sh:1]: ditto.

File: egs2/gigaspeech/asr2/run.sh [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+nclusters=1000
+
+src_lang=$(echo "${kmeans_feature}_km${nclusters}" | tr "/" "_")
+tgt_lang=en
+
+train_set="train"
+train_dev="dev"
+test_sets="dev test"
+
+asr_config=conf/tuning/train_discrete_asr_e_branchformer1_e12.yaml
```
Comment (simpleoier) [inline egs2/gigaspeech/asr2/run.sh:20]: Which config did you use actually?

File: espnet2/asr/encoder/e_branchformer_encoder.py
Comment (simpleoier) [inline espnet2/asr/encoder/e_branchformer_encoder.py:1]: can you add `Conv1d1` definition in `espnet/nets/pytorch_backend/transformer/subsampling.py`?

**Review by simpleoier — COMMENTED** [review]
Thanks for your efforts! The results are really cool!
I left some comments. Please reflect them.

**Comment by yichen14** [comment]
> Thanks for your efforts! The results are really cool! I left some comments. Please reflect them.

Hi Xuankai, thank you for the review! I've resolved those comments.

File: egs2/chime4/asr2/README.md [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+|decode_asr_asr_model_valid.acc.ave/dt05_simu_beamformit_2mics|1640|27120|89.8|8.4|1.7|0.7|10.8|61.8|
+|decode_asr_asr_model_valid.acc.ave/dt05_simu_beamformit_5mics|1640|27120|93.3|5.7|1.0|0.4|7.1|54.9|
+|decode_asr_asr_model_valid.acc.ave/dt05_simu_isolated_1ch_track|1640|27120|84.4|12.8|2.8|0.8|16.5|67.1|
+|decode_asr_asr_model_valid.acc.ave/et05_real_beamformit_2mics|1320|21409|90.5|8.2|1.3|0.6|10.1|66.7|
+|decode_asr_asr_model_valid.acc.ave/et05_real_beamformit_5mics|1320|21409|92.8|6.3|0.9|0.5|7.7|58.0|
+|decode_asr_asr_model_valid.acc.ave/et05_real_isolated_1ch_track|1320|21409|88.3|10.0|1.7|0.8|12.5|71.5|
+|decode_asr_asr_model_valid.acc.ave/et05_simu_beamformit_2mics|1320|21416|89.0|9.2|1.8|0.9|11.8|66.4|
+|decode_asr_asr_model_valid.acc.ave/et05_simu_beamformit_5mics|1320|21416|92.7|6.4|0.9|0.7|8.0|61.0|
+|decode_asr_asr_model_valid.acc.ave/et05_simu_isolated_1ch_track|1320|21416|83.3|13.2|3.5|1.1|17.8|69.4|
+|decode_asr_asr_model_valid.acc.best/dt05_real_beamformit_2mics|1640|27119|89.5|9.2|1.3|0.9|11.4|60.2|
```
Comment (simpleoier) [inline egs2/chime4/asr2/README.md:51]: Just keep the results of `valid.acc.ave`. You can delete `valid.acc.best` results.

File: egs2/chime4/asr2/conf/tuning/train_discrete_asr_e_branchformer_e12_mlp1024_linear1024_macaron_lr1e-4_warmup25k_conv1d1.yaml [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    rel_pos_type: latest
+    cgmlp_linear_units: 1024
+    cgmlp_conv_kernel: 31
+    use_linear_after_conv: false
+    gate_activation: identity
+    num_blocks: 12
+    dropout_rate: 0.1
+    positional_dropout_rate: 0.1
+    attention_dropout_rate: 0.1
+    input_layer: conv1d1  # using conv1d1 to ensure a proper input length in decode stage
```
Comment (simpleoier) [inline egs2/chime4/asr2/conf/tuning/train_discrete_asr_e_branchformer_e12_mlp1024_linear1024_macaron_lr1e-4_warmup25k_conv1d1.yaml:22]: Please be more specific, e.g. to ensure decoder input length > output length.

File: egs2/gigaspeech/asr2/README.md [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+### WER
+
+- ASR config: conf/tuning/train_discrete_asr_e_branchformer1_e12_lr.yaml
+- Decode config: conf/decode_ctc0.3.yaml
+- Model link: https://huggingface.co/espnet/eason_gigaspeech_train_asr2_e_branchformer12_lr_raw_wavlm_large_21_km1000
+
+|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
+|---|---|---|---|---|---|---|---|---|
+|decode_ctc0.3_asr_model_valid.acc.ave/test|19930|392325|89.0|6.9|4.1|1.9|12.9|73.4|
+|decode_ctc0.3_asr_model_valid.acc.best/test|19930|392325|88.8|6.8|4.3|1.9|13.1|74.5|
```
Comment (simpleoier) [inline egs2/gigaspeech/asr2/README.md:26]: Please keep only the `valid.acc.ave` result.
The `dev` result is missing.

File: egs2/gigaspeech/asr2/conf/tuning/train_discrete_asr_e_branchformer1_e12_lr.yaml
Comment (simpleoier) [inline egs2/gigaspeech/asr2/conf/tuning/train_discrete_asr_e_branchformer1_e12_lr.yaml:1]: Can you update the filename? The `lr` in the end is not meaningful.

File: egs2/gigaspeech/asr2/run.sh [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+nclusters=1000
+
+src_lang=$(echo "${kmeans_feature}_km${nclusters}" | tr "/" "_")
+tgt_lang=en
+
+train_set="train"
+train_dev="dev"
+test_sets="dev test"
+
+asr_config=conf/tuning/train_discrete_asr_e_branchformer1_e12_lr.yaml
```
Comment (simpleoier) [inline egs2/gigaspeech/asr2/run.sh:20]: Please update the config name here

File: espnet/nets/pytorch_backend/transformer/subsampling.py
Comment (simpleoier) [inline espnet/nets/pytorch_backend/transformer/subsampling.py:1]: Please add corresponding statement in `check_short_utt` for Conv1dSubsampling1 at line 33.

**Review by simpleoier — COMMENTED** [review]
Thanks! Can you address the following comments?

File: espnet/nets/pytorch_backend/transformer/subsampling.py [hunk 16]
```diff
@@ -30,6 +30,8 @@ def __init__(self, message, actual_size, limit):
 
 def check_short_utt(ins, size):
     """Check if the utterance is too short for subsampling."""
+    if isinstance(ins, Conv1dSubsampling1) and size < 1:
```
Comment (simpleoier) [inline espnet/nets/pytorch_backend/transformer/subsampling.py:33]: I think it is also `5` here.

Comment (yichen14) [reply espnet/nets/pytorch_backend/transformer/subsampling.py:33]: ok, got it

**Review by sw005320 — APPROVED** [review]
Great job!
Thanks

**Review by simpleoier — APPROVED** [review]
LGTM!
It's ready after fixing the CI errors.

**Comment by sw005320** [comment]
@yichen14, please fix https://github.com/espnet/espnet/actions/runs/6158168923/job/16807750493?pr=5427#step:7:100515
We're almost there.

**Comment by sw005320** [comment]
Hi @yichen14, this is a reminder.

**Comment by yichen14** [comment]
Sorry I missed previous comment, I'll fix it today

**Comment by ftshijt** [comment]
Thanks for your contribution!

**Pull request merged**